sycl/include/CL/sycl/half_type.hpp

```diff
@@ -86,22 +88,132 @@ inline float cast_if_host_half(half_impl::half val) {
 } // namespace sycl
 } // namespace cl
 
+#ifdef __SYCL_DEVICE_ONLY__
+using half = _Float16;
+#else
+using half = cl::sycl::detail::half_impl::half;
+#endif
+
+// Partial specialization of some functions in namespace `std`
 namespace std {
 
-template <> struct hash<cl::sycl::detail::half_impl::half> {
-  size_t operator()(cl::sycl::detail::half_impl::half const &key) const
-      noexcept {
-    return hash<uint16_t>()(key.Buf);
+#ifdef __SYCL_DEVICE_ONLY__
+// `constexpr` could work because the implicit conversion from `float` to
+// `_Float16` can be `constexpr`.
+#define CONSTEXPR_QUALIFIER constexpr
+#else
+// The qualifier is `const` instead of `constexpr` that is original to be
+// because the constructor is not `constexpr` function.
+#define CONSTEXPR_QUALIFIER const
+#endif
+
+// Partial specialization of `std::hash<cl::sycl::half>`
+template <> struct hash<half> {
+  size_t operator()(half const &Key) const noexcept {
+    return hash<uint16_t>()(reinterpret_cast<const uint16_t &>(Key));
   }
 };
 
-} // namespace std
+// Partial specialization of `std::numeric<cl::sycl::half>`
 
-#ifdef __SYCL_DEVICE_ONLY__
-using half = _Float16;
-#else
-using half = cl::sycl::detail::half_impl::half;
-#endif
+// All following values are either calculated based on description of each
+// function/value on https://en.cppreference.com/w/cpp/types/numeric_limits, or
+// cl_platform.h.
+#define HLF_MIN 6.103515625e-05F
+
+#define HLF_MAX 65504.0F
+
+#define HLF_MAX_10_EXP 4
+
+#define HLF_MAX_EXP 16
+
+#define HLF_MIN_10_EXP -4
+
+#define HLF_MIN_EXP -13
+
+#define HLF_MANT_DIG 11
+
+#define HLF_DIG 3
+
+#define HLF_DECIMAL_DIG 5
+
+#define HLF_EPSILON 9.765625e-04F
+
+template <> struct numeric_limits<half> {
+  static constexpr const bool is_specialized = true;
+
+  static constexpr const bool is_signed = true;
+
+  static constexpr const bool is_integer = false;
+
+  static constexpr const bool is_exact = false;
+
+  static constexpr const bool has_infinity = true;
+
+  static constexpr const bool has_quiet_NaN = true;
+
+  static constexpr const bool has_signaling_NaN = true;
+
+  static constexpr const float_denorm_style has_denorm = denorm_present;
+
+  static constexpr const bool has_denorm_loss = false;
+
+  static constexpr const bool tinyness_before = false;
+
+  static constexpr const bool traps = false;
+
+  static constexpr const int max_exponent10 = HLF_MAX_10_EXP;
```

Comment:
The constant could be just be defined here without macro.

Reply:
Explained above.